feat: Add comprehensive PnL tracking with Relay API integration

aldin4u opened the pull request:
- Implement PnL calculation using Relay API for accurate buy/sell tracking
- Add PnLStats component with 4-column layout (Balance USD, Bought, Sold, Floating PnL, Total PnL)
- Add TokenPnLCell component for portfolio list PnL display
- Create useTokenPnL hook for fetching and calculating PnL metrics
- Add auto-refresh (30s) and manual refresh support for PnL data
- Integrate PnL display in Buy and Sell components
- Add skeleton loading animations for PnL components
- Update My Holdings layout with improved column spacing and sort indicators
- Add Relay API service for fetching transaction data
- Add wallet transactions API integration
- Implement PnL utilities with comprehensive test coverage

## Description
<!--- Provide a general summary of your changes in the Title above -->
-

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
-

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Four-column P&L summary widget, per-token unrealised PnL cells, and a "Hide <$0.5" toggle in token list; sortable headers for Price, Balance, and PnL.

* **Improvements**
  * PnL wired to wallet transactions with periodic auto-refresh and on-demand refetch; improved loading UX with row and P&L skeletons; consistent color/formatting across Buy, Sell, and portfolio.

* **Tests**
  * Expanded mocks and coverage for PnL, wallet transactions, search, and portfolio flows.

* **Chores**
  * UI dependency bumped.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

aldin4u committed:
feat: Add PnL tracking with Relay API integration

- Implement PnL calculation using Relay API for accurate buy/sell tracking
- Add PnLStats component with 4-column layout
- Add TokenPnLCell component for portfolio list PnL display
- Create useTokenPnL hook for fetching and calculating PnL metrics
- Add auto-refresh (30s) and manual refresh support for PnL data
- Integrate PnL display in Buy and Sell components
- Add skeleton loading animations for PnL components
- Update My Holdings layout with improved column spacing and sort indicators
- Add Relay API service for fetching transaction data
- Add wallet transactions API integration
- Implement PnL utilities with test coverage

Refactored PnL utilities to comply with linting rules:
- Replaced for-of loops with Promise.all and array methods
- Eliminated continue statements using filter/map
- Added explicit return types and fixed hook dependencies

aldin4u committed:
fix: Resolve lint errors in PnL utilities and hooks

aldin4u committed:
fix: Remove TradingViewChart import to fix build error

aldin4u committed:
fix: Resolve infinite loop in useTokenPnL hook

Destructured props and updated useEffect dependency array to use specific
stable values (token.contract, walletAddress, etc.) instead of the unstable
props object. This prevents infinite re-rendering loops caused by
TokenPnLCell creating a new props object on every render. Also fixed a
TypeScript error by using token.contract instead of token.address.

aldin4u committed:
feat: Add 'Hide Small Amounts' toggle and fix test suite

- Move 'Hide <$0.5' toggle from PortfolioTokenList to Search.tsx header
- Toggle appears next to 'My Holdings' text in Sell mode
- Filters tokens with balance < $0.5 when enabled
- Fix test failures by adding complete RTK Query API mocks
- Add TestWrapper to Sell.test.tsx, HomeScreen.test.tsx, AppWrapper.test.tsx
- Add mocks for useTokenPnL, pillarXApiWalletTransactions, useRelayBuy, useRemoteConfig
- Update test snapshots for all modified components

Test Results:
- 946/951 tests passing (99.5% pass rate)
- Fixed 68 test failures across 4 test files
- UI manually verified and working correctly

aldin4u committed:
fix: correct PnL box positioning in Buy and resolve test failures

- Move PnL Stats in Buy.tsx to appear after buy button (matching Sell layout)
- Remove debug console.log from Sell.tsx
- Update @mui/joy to 5.0.0-beta.48 to resolve @mui/base deprecation
- Add eslint override to force v8.57.0 across all dependencies
- Fix Search.test.tsx Redux context errors by removing Provider and using proper mocks
- Update test snapshots for PortfolioTokenList

aldin4u committed:
fix: update test snapshots for Buy, PortfolioTokenList and TokenGraphColumn

aldin4u committed:
fix: resolve PnL discrepancy, formatting issues, and test snapshots

- Buy.tsx: Use portfolioToken for PnL calculation to match Sell.tsx logic
- PnLStats.tsx: Fix negative value formatting for USD and percentage
- CardSwap.test.tsx: Update snapshot to reflect removed button
- Buy.test.tsx & Sell.test.tsx: Update snapshots for PnL formatting changes

aldin4u committed:
fix: resolve lint errors in useTokenPnL.ts

- Remove extra space in refetch definition
- Remove unused transactionsData variable
- Suppress exhaustive-deps warning for intentional omission

aldin4u committed:
fix: resolve runtime error and lint issues manually

- Buy.tsx: Fix runtime error by destructuring transactionsData
- PnLStats.tsx: Fix nested ternary lint error
- Search.test.tsx: Remove unused imports
- useTokenPnL.ts: Fix exhaustive-deps warning by adding token dependency
- Buy.tsx: Fix formatting issues using prettier

aldin4u committed:
fix: resolve infinite loop in useTokenPnL and improve PnL refresh animation

- Fix infinite loop caused by unstable 'token' dependency in useEffect
- Fix auto-refresh effect dependencies (props -> stable primitives)
- Add 100ms minimum loading duration for skeleton animation visibility
- Add 5s timeout to all Relay API requests to prevent hangs
- Fix PnLStats visibility in Buy/Sell components (add isTransactionsLoading)
- Add isRefreshing prop to TokenPnLCell for refresh coordination
- Move refetchPnL to start of refreshBuyIntent for immediate refresh
- Fix TokenPnLCell imports and property names (PortfolioToken)

aldin4u committed:
fix: correct PnL calculation for Relay transactions with nested metadata

- Fix metadata extraction to check both req.metadata and req.data.metadata
- Update RelayRequest TypeScript interface to include nested metadata field
- Add validation to prevent absurdly large USDC amounts (>T)
- Add validation to prevent invalid execution prices (>M or <=0)
- Add console warnings for skipped trades with suspicious values

This fixes the issue where CAKE token showed execution price of B
instead of the correct .67 due to metadata being nested in req.data.metadata

aldin4u committed:
fix: return null PnL for sell-only tokens to display '-'

RanaBug committed:
clean random test files and fix some unrelated changes

RanaBug commented on src/utils/pnl.ts:
```diff
@@ -0,0 +1,547 @@
+    .map(({ txHash, relayReq }) => {
+      if (!relayReq) return null;
+
+      // Check if USDC is involved in the Relay transaction (via stateChanges)
```
@aldin4u why only 4 USDC addresses there?

RanaBug replied:
Added the stablecurrency constant list instead

RanaBug commented on src/utils/pnl.ts:
```diff
@@ -0,0 +1,547 @@
+
+  // Known USDC addresses for matching quote currency
+  const USDC_ADDRESSES = [
+    '0x833589fcd6edb6e08f4c7c32d4f71b54bda02913', // Base USDC
```
We are repeating the USDC addresses there... and they are more here than above. We need one list to refer to each time

RanaBug commented on src/services/relayApi.ts:
```diff
@@ -0,0 +1,180 @@
+import axios from 'axios';
+
+export interface RelayRequest {
```
Surely there is a Relay type we could import from their SDK rather than copying all params in the type here.

RanaBug replied:
Relay does not seem to provide an exportable type for this relayreuqest

RanaBug commented on src/services/relayApi.ts:
```diff
@@ -0,0 +1,180 @@
+
+const RELAY_API_URL = 'https://api.relay.link';
+
+export const fetchRelayRequests = async (
```
We use RTQ query rather than axios for consistency

RanaBug replied:
Done

RanaBug commented on src/services/pillarXApiWalletPortfolio.ts:
```diff
@@ -190,6 +190,7 @@ export const pillarXApiWalletPortfolio = createApi({
           },
         };
       },
+      keepUnusedDataFor: 300, // Cache for 5 minutes
```
Why this change?

RanaBug replied:
Removed

RanaBug commented on src/hooks/useTokenPnL.ts:
```diff
@@ -0,0 +1,317 @@
+  });
+
+  // eslint-disable-next-line no-console
+  console.log('[useTokenPnL] Hook render', {
```
Do we want to keep this log in?

RanaBug replied:
Removed

RanaBug commented on src/apps/pulse/components/Sell/Sell.tsx:
```diff
@@ -74,6 +79,58 @@ const Sell = (props: SellProps) => {
+            ...token,
+            contract: token.address || '',
+            id: token.symbol,
+            blockchain: 'ethereum',
```
Why is it static on "ethereum" chain?

RanaBug replied:
Fixed with token blockchain

RanaBug commented on src/apps/pulse/components/Sell/Sell.tsx:
```diff
@@ -74,6 +79,58 @@ const Sell = (props: SellProps) => {
+            price: portfolioToken.price || 0,
+            decimals: token.decimals || 18,
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          } as any,
```
too much "any" type - there is a reason why we avoid "any" and it is flagged as an eslint error that has been disable in this line and the next one: this is to avoid errors.

RanaBug replied:
Fixed

RanaBug commented on src/apps/pulse/components/Search/TokenPnLCell.tsx:
```diff
@@ -0,0 +1,93 @@
+            balance: token.balance, // Added balance
+            price: token.price, // Added price
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          } as any, // Cast to any
```
"any" again, we need to avoid "any" as a type

RanaBug replied:
Fixed

RanaBug commented on src/apps/pulse/components/Search/PortfolioTokenList.tsx:
```diff
@@ -20,24 +21,102 @@ import { STABLE_CURRENCIES } from '../../constants/tokens';
+import { useGetWalletTransactionsQuery } from '../../../../services/pillarXApiWalletTransactions';
+
+// Sort indicator component
+const SortIndicator = ({
```
Needs to be in its own file as this is a component

RanaBug replied:
Fixed

RanaBug commented on src/apps/pulse/components/Search/PortfolioTokenList.tsx:
```diff
@@ -20,24 +21,102 @@ import { STABLE_CURRENCIES } from '../../constants/tokens';
 }
 
+// Skeleton loader component
+const SkeletonTokenRow = () => (
```
Needs to be in its own file as this is a component

RanaBug replied:
Fixed

RanaBug commented on src/apps/pulse/components/PnLStats/PnLStats.tsx:
```diff
@@ -0,0 +1,215 @@
+  isLoading: boolean;
+}
+
+const StatGroup = ({
```
Needs to be in its own file as this is a component

RanaBug replied:
Fixed

RanaBug commented on src/apps/pulse/components/PnLStats/PnLStats.tsx:
```diff
@@ -0,0 +1,215 @@
+);
+
+// Skeleton loader for PnL stats
+const PnLStatsSkeleton = () => (
```
Needs to be in its own file as this is a component

RanaBug replied:
Fixed

RanaBug commented on src/apps/pulse/components/Buy/Buy.tsx:
```diff
@@ -165,6 +167,50 @@ export default function Buy(props: BuyProps) {
+            ...token,
+            contract: token.address || '',
+            id: token.symbol,
+            blockchain: 'ethereum',
```
Why only "ethereum" here? And not token.chain?

RanaBug replied:
Fixed with token blockchain

RanaBug commented on src/apps/pulse/components/Buy/Buy.tsx:
```diff
@@ -165,6 +167,50 @@ export default function Buy(props: BuyProps) {
+            price: portfolioToken.price || 0,
+            decimals: token.decimals || 18,
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          } as any,
```
Need to be a type, not any in any....

RanaBug replied:
Fixed

RanaBug commented on src/apps/pulse/components/Buy/Buy.tsx:
```diff
@@ -784,7 +836,8 @@ export default function Buy(props: BuyProps) {
           expressIntentResponse={
-            USE_RELAY_BUY ? buyOffer : expressIntentResponse
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            USE_RELAY_BUY ? (buyOffer as any) : expressIntentResponse
```
Not any...

RanaBug replied:
Fixed

RanaBug committed:
lint fix

RanaBug commented on inspect_aave_portfolio.js:
```diff
@@ -0,0 +1,23 @@
+    console.log('  Contract:', aave?.asset?.contracts?.[0] || aave?.asset?.contract);
+    console.log('  Chain ID:', aave?.asset?.chain_id); // Check if this field exists
+    console.log('  Blockchain:', aave?.asset?.blockchain); // Or this
+})();
```
file has been deleted

RanaBug commented on inspect_full_aave.js:
```diff
@@ -0,0 +1,19 @@
+    const aave = assets.find(a => a.asset.symbol === 'AAVE');
+    console.log('Full AAVE Portfolio Item:');
+    console.log(JSON.stringify(aave, null, 2));
+})();
```
file has been deleted

RanaBug commented on test_portfolio_pnl.ts:
```diff
@@ -0,0 +1,166 @@
+                    relayValidatedCount++;
+
+                    // Analyze this transaction for USDC + token
+                    const txGroup = assetTxs.filter(tx => tx.hash === txHash);
```
file has been deleted

RanaBug commented on test_relay_pnl.ts:
```diff
@@ -0,0 +1,119 @@
+
+    // 2. Fetch Relay requests
+    console.log('2. Fetching Relay requests...');
+    const relayResponse = await fetch(`${RELAY_API_URL}/requests?user=${WALLET_ADDRESS}&status=filled`);
```
file has been deleted

RanaBug commented on test_relay_pnl.ts:
```diff
@@ -0,0 +1,119 @@
+    const tokensWithRelay: any[] = [];
+
+    tokenMap.forEach((tokenData: any, symbol: string) => {
+        const txHashes = Array.from(new Set(tokenData.transactions.map((tx: any) => tx.tx_hash.toLowerCase()))) as string[];
```
file has been deleted

RanaBug commented on verify_aave_pnl.js:
```diff
@@ -0,0 +1,144 @@
+        // So `isInbound` would be false.
+
+        // Let's check `pnl.ts` again carefully.
+    }
```
file has been deleted

RanaBug commented on verify_hybrid_pnl.js:
```diff
@@ -0,0 +1,214 @@
+            balance: cb.balance,
+            price: aaveAsset.price,
+            decimals: cb.decimals
+        }))[0]; // Ethereum contract
```
file has been deleted

RanaBug committed:
minor fixes

RanaBug committed:
Merge branch 'staging' into feature-pnl

RanaBug committed:
refresh button for pnl fix

RanaBug commented on src/hooks/useTokenPnL.ts:
```diff
@@ -0,0 +1,310 @@
+            },
+          }));
+
+        const relayResults = await Promise.all(
```
@IAmKio @aldin4u just a bit worried here as we are calling the relay api 20 times per token! Is there not a risk of being rate limited there?

aldin4u replied:
Don't think this is an issue since its user calling it and we are reducing the amount of calls by filtering. If the user gets rate limited, it would be great problem to address since they are trading so much :)

RanaBug committed:
timestamp fix

RanaBug commented on src/apps/pulse/components/Search/TokenPnLCell.tsx:
```diff
@@ -0,0 +1,52 @@
+            >
+                {pnl.unrealisedPnLPct >= 0 ? '+' : ''}
+                {pnl.unrealisedPnLPct.toFixed(2)}%
+            </p>
```
@coderabbitai Is this still valid CodeRabbit?

RanaBug replied:
The colour provides the cue that is is negative or positive and the percenatage

RanaBug commented on src/apps/token-atlas/components/TokenGraphColumn/TokenGraphColumn.tsx:
```diff
@@ -304,19 +303,29 @@ const TokenGraphColumn = ({
+                    '1D'
+        }
+        className="mt-4"
+      />
```
@coderabbitai Is this still valid CodeRabbit?

RanaBug commented on src/utils/pnl.ts:
```diff
@@ -0,0 +1,508 @@
+    if (trade.side === 'BUY') {
+      totalTokens += trade.amountToken;
+      totalCostUSDC += trade.amountQuoteUSDC;
+    } else {
+      // SELL
+      if (totalTokens === 0) return; // Selling without inventory (shouldn't happen in perfect history)
+
+      const wac = totalCostUSDC / totalTokens;
+      const costBasis = trade.amountToken * wac;
+
+      totalTokens -= trade.amountToken;
+      totalCostUSDC -= costBasis;
+
+      totalCostBasisSold += costBasis;
+      realisedPnLUSDC += trade.amountQuoteUSDC - costBasis;
+    }
```
@coderabbitai Is this still valid CodeRabbit?

RanaBug replied:
Fixed

RanaBug commented on src/utils/pnl.ts:
```diff
@@ -0,0 +1,508 @@
+      usdcAmount = 0;
+      if (relayReq.data?.inTxs) {
+        relayReq.data.inTxs.forEach((inTx) => {
+          if (inTx.stateChanges) {
+            inTx.stateChanges.forEach((stateChange) => {
+              const tokenAddr =
+                stateChange.change?.data?.tokenAddress?.toLowerCase();
+              const changeAddress = stateChange.address?.toLowerCase();
+
+              // Check if this is a USDC state change for the user
+              if (
+                tokenAddr &&
+                USDC_ADDRESSES.some(
+                  (addr) => addr.toLowerCase() === tokenAddr
+                ) &&
+                changeAddress === userAddress
+              ) {
+                hasUSDCInRelay = true;
+
+                // Extract amount from balanceDiff
+                const balanceDiffStr =
+                  (stateChange.change as { balanceDiff?: string })
+                    ?.balanceDiff || '0';
+                const balanceDiff = parseFloat(balanceDiffStr);
+
+                // For BUY, we expect negative USDC (spending)
+                // For SELL, we expect positive USDC (receiving)
+                if (side === 'BUY' && balanceDiff < 0) {
+                  usdcAmount += Math.abs(balanceDiff) / 1e6; // USDC has 6 decimals
+                } else if (side === 'SELL' && balanceDiff > 0) {
+                  usdcAmount += balanceDiff / 1e6;
+                }
+              }
+            });
+          }
+        });
+      }
```
@coderabbitai Is this still valid CodeRabbit?

RanaBug replied:
@aldin4u can you check here? Are we supposed here to only check relayReq.data.intTxs?

RanaBug commented on src/apps/pulse/components/Sell/Sell.tsx:
```diff
@@ -74,6 +79,54 @@ const Sell = (props: SellProps) => {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      } as any)
+      : null
+  );
```
@coderabbitai Is this still valid CodeRabbit?

RanaBug replied:
Not true, this has been updated and fixed already

RanaBug commented on src/hooks/useTokenPnL.ts:
```diff
@@ -0,0 +1,282 @@
+        }
+
+        // Calculate PnL from Relay data
+        const trades = calculatePnLFromRelay(relayRequests, {
+          address: tokenContract || '',
+          symbol: tokenSymbol || '',
+          decimals: tokenDecimals || 18,
+          chainId: chainId!,
+          price: tokenPrice,
+        });
```
@coderabbitai Is this still valid CodeRabbit?

RanaBug commented on src/apps/pulse/components/Buy/Buy.tsx:
```diff
@@ -163,6 +167,50 @@ export default function Buy(props: BuyProps) {
+  const {
+    pnl,
+    isLoading: isPnLLoading,
+    refetch: refetchPnL,
+  } = useTokenPnL(
+    token && accountAddress && portfolioToken
+      ? ({
+          token: {
+            ...token,
+            contract: token.address || '',
+            id: token.symbol,
+            blockchain: 'ethereum',
+            balance: portfolioToken.balance || 0,
+            price: portfolioToken.price || 0,
+            decimals: token.decimals || 18,
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          } as any,
+          transactionsData,
+          walletAddress: accountAddress,
+          chainId: token.chainId,
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        } as any)
+      : null
+  );
```
@coderabbitai Is this still valid CodeRabbit?

RanaBug commented on src/utils/pnl.ts:
```diff
@@ -0,0 +1,541 @@
+        if (symbol === 'USDC') {
+          if (isInbound) usdcChange += amount;
+          if (isOutbound) usdcChange -= amount;
+        }
```
@coderabbitai Is this still valid CodeRabbit?

RanaBug committed:
coderabbit fixes

RanaBug commented on src/utils/pnl.ts:
```diff
@@ -0,0 +1,545 @@
+      // Sanity check: USDC amount shouldn't exceed $1 trillion
+      if (amountUSDC > 1e12) {
+        console.warn(
+          `[calculatePnLFromRelay] Suspiciously large USDC amount for ${token.symbol}: $${amountUSDC.toLocaleString()}. Skipping trade ${req.id}`
+        );
+        return null;
+      }
+
+      // Sanity check: Execution price shouldn't exceed $1 million per token
+      if (execPrice > 1e6) {
+        console.warn(
+          `[calculatePnLFromRelay] Suspiciously high execution price for ${token.symbol}: $${execPrice.toLocaleString()}/token. Skipping trade ${req.id}`
+        );
+        return null;
+      }
+
+      // Sanity check: Execution price shouldn't be negative or zero
+      if (execPrice <= 0) {
+        console.warn(
+          `[calculatePnLFromRelay] Invalid execution price for ${token.symbol}: $${execPrice}. Skipping trade ${req.id}`
+        );
+        return null;
+      }
+
+      return {
+        side,
+        txHash: req.id,
+        timestamp,
+        amountToken,
+        amountQuoteUSDC: amountUSDC,
+        execPriceUSD: execPrice,
+        feesUSD: 0,
+        tokenAddress: token.address,
+        tokenSymbol: token.symbol,
+      };
+    })
+    .filter((trade) => trade !== null) as ReconstructedTrade[];
+
+  return trades.sort((a, b) => a.timestamp - b.timestamp);
+};
```
Fixed

RanaBug committed:
fix balance in PnL stats

aldin4u committed:
fix: prioritize state changes over metadata for PnL detection

- Changed trade detection to check state changes first before metadata
- Fixes issue where LINK sells were hidden in state changes while metadata showed USDC bridge
- Ensures all token movements are captured in complex multi-step transactions
- Verified fix detects previously missed LINK sell (0.039147 LINK for $0.01)

RanaBug committed:
lint and unit tests fix

RanaBug commented on src/utils/pnl.ts:
```diff
@@ -0,0 +1,548 @@
+
+      // For USDC amount, extract it from Relay stateChanges
+      usdcAmount = 0;
+      if (relayReq.data?.inTxs) {
+        relayReq.data.inTxs.forEach((inTx) => {
+          if (inTx.stateChanges) {
+            inTx.stateChanges.forEach((stateChange) => {
+              const tokenAddr =
+                stateChange.change?.data?.tokenAddress?.toLowerCase();
+              const changeAddress = stateChange.address?.toLowerCase();
+
+              // Check if this is a USDC state change for the user
+              if (
+                tokenAddr &&
+                USDC_ADDRESSES.some((addr: string) => addr === tokenAddr) &&
+                changeAddress === userAddress
+              ) {
+                hasUSDCInRelay = true;
+
+                // Extract amount from balanceDiff
+                const balanceDiffStr =
+                  (stateChange.change as { balanceDiff?: string })
+                    ?.balanceDiff || '0';
+                const balanceDiff = parseFloat(balanceDiffStr);
+
+                // For BUY, we expect negative USDC (spending)
+                // For SELL, we expect positive USDC (receiving)
+                if (side === 'BUY' && balanceDiff < 0) {
+                  usdcAmount += Math.abs(balanceDiff) / 1e6; // USDC has 6 decimals
+                } else if (side === 'SELL' && balanceDiff > 0) {
+                  usdcAmount += balanceDiff / 1e6;
+                }
+              }
+            });
+          }
+        });
+      }
```
To be ignored @aldin4u

vignesha22 reviewed (approved):
LGTM a lot of formatting from prettier on most of the files
